### PR TITLE
Make the metric in the buckets_path parameter optional

### LINKED ISSUE #72983: top_metrics and bucket_script params should automagically work
**Elasticsearch version** (`bin/elasticsearch --version`): 7.12

**OS version** (`uname -a` if on a Unix-like system): Running Elastic Cloud.

**Description of the problem including expected versus actual behavior**:

Combining `bucket_script` with `top_metrics` should just work.

In the tutorials and/or docs on elastic.co the `bucket_script params` are often referred to child / grandparents syntax. However that does not work.

```
current-price>top
current-price>top>metrics
current-price>top>metrics>price
```

or (change everything to dot-walking)

```
current-price.top.metrics.price
```

I get either that error

```
"reason" : "buckets_path must reference either a number value or a single value numeric metric aggregation, but [current-price] contains multiple values. Please specify which to use."
```

or

```
"reason" : "Validation Failed: 1: No aggregation [top] found for path [current-price>top];"
```

or (when dot walking)

```
"reason" : "path not supported for [current-price]: [top]"
```

**How should it work**:

1. The `top_metrics` looks like this:

```json
"current-price" : {
    "top" : [
      {
        "sort" : [
          "+51361822-08-28T19:06:40.000Z"
        ],
        "metrics" : {
          "price" : 4.94372760037313
        }
      }
    ]
 }
```

2. Elasticsearch should *magically* detect that the *current-price* top only contains *one* valid value.
3. I should be able to write `buckets_path: { "currentprice":"current-price"}` and *not* `current-price.price`.

**Working version**:

<details>
  <summary>Query</summary>

```json
{
  "query": {
    "range": {
      "@timestamp": {
        "gte": "now-1y/d",
        "lte": "now/d"
      }
    }
  },
  "size": 0,
  "aggs": {
    "terms_agg": {
      "terms": {
        "field": "pair"
    },
    "aggs":{
      "sum-of-expenses":{
        "sum": {
          "script": """
            if (doc['type'].contains('buy')) {
              return doc['cost']
            }
          """
          }
        },
        "amount-of-coin":{
          "sum": {
            "script": """
              if (doc['type'].contains('buy')) {
                return doc['vol']
              }
            """
          }
        },
        "current-price":{
         "top_metrics":{
            "metrics": { "field": "price" },
            "sort": { "@timestamp": "desc" },
            "size": 1
            }
        },
        "value-of-depot": {
          "bucket_script": {
            "buckets_path": {
              "coins": "amount-of-coin",
              "currentprice": "current-price.price"
            },
            "script": "params.coins * params.currentprice"
          }
        }
      }
    }
   }
}
```
</details>

<details>
  <summary>Answer</summary>

```json
{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 435,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "terms_agg" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [
        {
          "key" : "DOTEUR",
          "doc_count" : 221,
          "current-price" : {
            "top" : [
              {
                "sort" : [
                  "+51363695-02-21T20:00:00.000Z"
                ],
                "metrics" : {
                  "price" : 32.72653956292378
                }
              }
            ]
          },
          "amount-of-coin" : {
            "value" : 14.2345928
          },
          "sum-of-expenses" : {
            "value" : 402.59811
          },
          "value-of-depot" : {
            "value" : 465.84896443131004
          }
        },
        {
          "key" : "KAVAEUR",
          "doc_count" : 214,
          "current-price" : {
            "top" : [
              {
                "sort" : [
                  "+51363694-07-16T22:13:20.000Z"
                ],
                "metrics" : {
                  "price" : 5.091578288250437
                }
              }
            ]
          },
          "amount-of-coin" : {
            "value" : 9.0
          },
          "sum-of-expenses" : {
            "value" : 49.95
          },
          "value-of-depot" : {
            "value" : 45.824204594253935
          }
        }
      ]
    }
  }
}
```
</details>

### PR DESCRIPTION
**Pull request by salvatore-campagna**
With this change the metric field name becomes optional if the
`bukets_path` is pointing to a multi-value aggregation with a single
metric field. Normally the full path would be required including
the aggregation name followed by the metric field.

If the metric is no specified in the path and the multi-value
aggregation computes more than one value an error is thrown.

The old notation is still supported for backward compatibility in case
the full path is specified and the target multi-value aggregation
computes a single value.

Resolves #72983.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
@not-napoleon According to my understanding those aggregations are the only ones where we can apply this kind of "shortcut" for the path. That is because they return an array as a result and specifying a metric might not be necessary. For those which return an object (like for instance extended stats) we still need the full path.

**Review by not-napoleon — CHANGES_REQUESTED** [review]
The yaml tests are good, but I'd like to see some java tests as well.  As noted, I'd like to confirm that putting a nonsense name in the last position fails.  I'd also like to test that if your percentile has e.g. 99 and you ask for 95, that fails correctly.  I think it will, but let's make sure we have a test.

Also, just as a note, if instead of "addresses issue #..." you say "resolves #...", github will automatically close the issue when the PR is merged.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractInternalHDRPercentiles.java [hunk 1]
```diff
@@ -78,7 +78,14 @@ protected void doWriteTo(StreamOutput out) throws IOException {
 
     @Override
     public double value(String name) {
-        return value(Double.parseDouble(name));
+        try {
+            return value(Double.parseDouble(name));
```
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractInternalHDRPercentiles.java:82]: Correct...I didn't realise we can just compare the name (parameter) to the aggregation name 🤦

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Review by not-napoleon — APPROVED** [review]
LGTM, thanks for addressing the changes.

**Pull request merged**